packages/airnode-utilities/src/evm/gas-prices/gas-oracle.ts

```diff
@@ -0,0 +1,221 @@
+      latestPercentileGasPrice.percentileGasPrice,
+      referencePercentileGasPrice.percentileGasPrice,
+      maxDeviationMultiplier
+    );
+
+    // Return the percentile for the latest block if within the limit
+    if (isWithinDeviationLimit) {
+      return latestPercentileGasPrice.percentileGasPrice;
+    }
+
+    throw new Error(
+      `Latest block percentileGasPrice exceeds the max deviation multiplier limit set to (${maxDeviationMultiplier}%).`
+    );
+  } else {
+    throw new Error(`Unable to get enough blocks or transactions to calculate percentileGasPrice.`);
+  }
+};
+
+export const attemptGasOracleStrategy = async (
+  provider: ethers.providers.StaticJsonRpcProvider,
+  gasOracleConfig: GasPriceOracleStrategy
+) => {
+  switch (gasOracleConfig.gasPriceStrategy) {
+    case 'latestBlockPercentileGasPrice':
+      return await fetchLatestBlockPercentileGasPrice(provider, gasOracleConfig);
+    case 'providerRecommendedGasPrice':
+      return await fetchProviderRecommendedGasPrice(provider, gasOracleConfig);
+    case 'constantGasPrice':
+      return fetchConstantGasPrice(gasOracleConfig);
+  }
+};
+
+// Get gas price based on gas price oracle strategies
+export const getGasPrice = async (
+  provider: ethers.providers.StaticJsonRpcProvider,
+  gasPriceOracleConfig: GasPriceOracleConfig
+): Promise<LogsData<ethers.BigNumber>> => {
+  const logs: PendingLog[] = [];
+
+  // Attempt gas oracle strategies in order
+  for (const strategy of gasPriceOracleConfig) {
+    const goAttemptGasOraclePriceStrategy = await go(() => attemptGasOracleStrategy(provider, strategy), {
+      retries: 0,
+    });
+
+    // Continue to the next strategy attempt if the current fails
+    if (!goAttemptGasOraclePriceStrategy.success) {
+      logs.push(
+        logger.pend(
+          'ERROR',
+          `Strategy (${strategy.gasPriceStrategy}) failed to return a gas price. Error: ${goAttemptGasOraclePriceStrategy.error.message}.`
+        )
+      );
+      continue;
+    }
+
+    // Otherwise return the gas price
+    logs.push(
+      logger.pend(
+        'INFO',
+        `Strategy (${strategy.gasPriceStrategy}) gas price set to ${ethers.utils.formatUnits(
+          goAttemptGasOraclePriceStrategy.data,
+          'gwei'
+        )} gwei.`
+      )
+    );
+    return [logs, goAttemptGasOraclePriceStrategy.data];
+  }
+
+  // Return the constant strategy gas price if all other strategies fail
+  const constantGasPriceConfig = gasPriceOracleConfig.find(
+    (strategy) => strategy.gasPriceStrategy === 'constantGasPrice'
+  ) as ConstantGasPriceStrategy;
+  const constantGasPrice = fetchConstantGasPrice(constantGasPriceConfig);
+  logs.push(
+    logger.pend(
+      'INFO',
+      `All oracle strategies failed to return a gas price, returning constant gas price set to ${ethers.utils.formatUnits(
+        constantGasPrice,
+        'gwei'
+      )} gwei.`
+    )
+  );
+
+  return [logs, constantGasPrice];
```

Comment:
Are we even going to get here? Isn't the `constantGasPrice` strategy mandatory and so will be processed in the code above among other strategies?

Reply:
You're right, but I had to add that as without it typescript shows an error that '`Not all code paths return a value`'. 

Perhaps I need to refactor the logic a bit and handle the `constantGasPrice` strategy separately if you see this as an issue?

Reply:
I see. It's not an issue in terms that it will cause problems but the code is a bit misleading for whoever reads it.

What about changing the validation so that the `constantGasPrice` strategy must be the last one in the array (it doesn't make sense to have any other strategies behind this one, they won't be processed anyway). Then, looping through all the ones in front of it. And then, processing `constantGasPrice`.